mpmath/matrices/matrices.py

```diff
@@ -993,3 +993,26 @@ def mnorm(ctx, A, p=1):
             return max(ctx.fsum((A[i,j] for j in range(n)), absolute=1) for i in range(m))
         else:
             raise NotImplementedError("matrix p-norm for arbitrary p")
+        
+    def kron(ctx, *args):
+        """
+        Calculates the tensor products of 2 or more matrices, similar to numpy.kron
+        """
+        new_rows: int = 1
+        new_cols: int = 1
+        for m in args:
+            new_rows *= m.rows
+            new_cols *= m.cols
+        res = ctx.ones(new_rows, new_cols)
+        for i in range(new_rows):
+            for j in range(new_cols):
+                partition_rows: int = 1
+                partition_cols: int = 1
+                for m in args:
+                    partition_rows *= m.rows
+                    partition_cols *= m.cols
+                    res[i, j] *= m[
+                            int(i * partition_rows / new_rows) % m.rows, 
+                            int(j * partition_cols / new_cols) % m.cols
+                        ]
```

Comment:
Why not use same two-arguments form as numpy.kron?  As TP is associative, we could use reduce to compute more general case.

Given ``def kron(ctx, a, b):``,
```suggestion
                res[i, j] = a[i // b.rows, j // b.cols]*b[i % b.rows, j % b.cols]
```

Reply:
I'm not sure what would be gained from doing this, and it would make things significantly less efficient for calculations involving more than two inputs since it would involve allocating and re-allocating memory for successively larger intermediate matrices

Reply:
> I'm not sure what would be gained from doing this

Simple API (compatible wrt numpy), obvious code.  You also could initialize res as an empty matrix.

> significantly less efficient for calculations involving more than two inputs

How much gain we well have, could argue this?  And/Or provide benchmarks?  I would expect, that multiplication cost will be dominant, esp. for the mp context.

```
In [21]: %timeit -r11 r=mp.kron(a,a,a)
3.98 ms ± 3.72 µs per loop (mean ± std. dev. of 11 runs, 100 loops each)

In [22]: %timeit -r11 r=reduce(mp.kron, [a, a, a])
3.55 ms ± 12.2 µs per loop (mean ± std. dev. of 11 runs, 100 loops each)

In [23]: a = fp.matrix([[1, 2], [3, 4]])

In [24]: %timeit -r11 r=reduce(fp.kron, [a, a, a])
1.53 ms ± 16.5 µs per loop (mean ± std. dev. of 11 runs, 1,000 loops each)

In [25]: %timeit -r11 r=fp.kron(a,a,a)
1.71 ms ± 5.35 µs per loop (mean ± std. dev. of 11 runs, 1,000 loops each)

In [26]: def kron2(ctx, a, b):
    ...:     new_rows = a.rows*b.rows
    ...:     new_cols = a.cols*b.cols
    ...:     res = ctx.matrix(new_rows, new_cols)
    ...:     for i in range(new_rows):
    ...:         for j in range(new_cols):
    ...:             res[i, j] = a[i // b.rows, j // b.cols]*b[i % b.rows, j % b.cols]
    ...:     return res
    ...: 

In [27]: kron2fp = lambda a, b: kron2(fp, a, b)

In [28]: %timeit -r11 r=reduce(kron2fp, [a, a, a])
632 µs ± 1.66 µs per loop (mean ± std. dev. of 11 runs, 1,000 loops each)

In [29]: kron2mp = lambda a, b: kron2(mp, a, b)

In [30]: a = mp.matrix([[1, 2], [3, 4]])

In [31]: %timeit -r11 r=reduce(kron2mp, [a, a, a])
1.41 ms ± 12 µs per loop (mean ± std. dev. of 11 runs, 1,000 loops each)
```